packages/babel-core/src/transformation/file/options/option-manager.js

```diff
@@ -221,8 +221,19 @@ export default class OptionManager {
 
     // resolve presets
     if (opts.presets) {
-      this.mergePresets(opts.presets, dirname);
-      delete opts.presets;
+      // If we're in the "pass per preset" mode, we resolve the presets
+      // and keep them for further execution to calculate the options.
+      if (opts.passPerPresset) {
+        opts.presets = this.resolvePresets(opts.presets, dirname, (preset, presetLoc) => {
+          if (preset.plugins) {
+            preset.plugins = OptionManager.normalisePlugins(presetLoc, dirname, preset.plugins);
```

Comment:
@kittens: if we want to handle other options in presets (except `plugins`), then we could call `mergeOptions` here recursively, passing options "to merge into" as a parameter. So it'll be preset options, but not just top-level `this.options`, when it calls `merge(this.options, opts);` below.

It may look a bit weird though as `mergeOptions(preset, preset, ...);`. Or we can refactor `mergeOptions` since currently it does much more than actual merging. It does also normalization, e.g. `preset.plugins = OptionManager.normalisePlugins(...)` -- that exactly what we need. We don't wanna merge preset into itself, but we want to merge e.g. nested stuff in the presets like `envOpts` if they appear in the preset.

Or, better, to limit presets only to one nested level, and handle only small sub-set of options :smile: Let me know.